qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -116,77 +140,202 @@ def __init__(self, hls_config=None):
-                else:
-                    plugin_method = plugin_specifier
+        # The pass is recursive, as we may have annotated gates whose definitions
+        # consist of other annotated gates, whose definitions include for instance
+        # LinearFunctions. Note that in order to synthesize a controlled linear
+        # function, we must first fully synthesize the linear function, and then
+        # synthesize the circuit obtained by adding control logic.
+
+        # copy dag_op_nodes because we are modifying the DAG below
+        dag_op_nodes = dag.op_nodes()
 
-                # ToDo: similarly to UnitarySynthesis, we should pass additional parameters
-                #       e.g. coupling_map to the synthesis algorithm.
-                decomposition = plugin_method.run(node.op, **plugin_args)
+        for node in dag_op_nodes:
+            decomposition = self._recursively_handle_op(node.op)
 
-                # The synthesis methods that are not suited for the given higher-level-object
-                # will return None, in which case the next method in the list will be used.
-                if decomposition is not None:
-                    dag.substitute_node_with_dag(node, circuit_to_dag(decomposition))
-                    break
+            if not isinstance(decomposition, (QuantumCircuit, Operation)):
+                raise TranspilerError(f"HighLevelSynthesis was unable to synthesize {node.op}.")
+
+            if isinstance(decomposition, QuantumCircuit):
+                dag.substitute_node_with_dag(node, circuit_to_dag(decomposition))
+            elif isinstance(decomposition, Operation):
+                dag.substitute_node(node, decomposition)
 
         return dag
 
+    def _recursively_handle_op(self, op: Operation) -> Union[Operation, QuantumCircuit]:
+        """Recursively synthesizes a single operation.
+
+        The result can be either another operation or a quantum circuit.
+
+        Some examples when the result can be another operation:
+        Adding control to CX-gate results in CCX-gate,
+        Adding inverse to SGate results in SdgGate.
+
+        Some examples when the result can be a quantum circuit:
+        Synthesizing a LinearFunction produces a quantum circuit consisting of
+        CX-gates.
+
+        The function recursively handles operation's definition, if it exists.
+        """
+
+        # First, try to apply plugin mechanism
+        decomposition = self._synthesize_op_using_plugins(op)
+        if decomposition:
+            return decomposition
+
+        # Second, handle annotated operations
+        decomposition = self._synthesize_annotated_op(op)
+        if decomposition:
+            return decomposition
+
+        # Third, optionally recursively descend into op's definition if exists
+        if self.hls_config.recurse:
+            try:
+                # extract definition
+                definition = op.definition
```

Comment:
This part of the code was adapted from `UnrollCustomDefinitions`, however I understand that  `UnrollCustomDefinitions` runs at most once during transpilation, while in theory `HighLevelSynthesis` could be run multiple times (e.g., as part of the optimization loop, but this does not happen right now). I did add an argument `recurse` which would prevent examining any definitions. And yes HLS needs to support recursion (not doing so was an omission in the original implementation).

Reply:
Yeah, I'm worried about the performance of this too since we'll potentially be doing it on every node. We need to think a bit about our strategy here. Does it make sense to try and limit the scope here to just what is needed for annotated operations? 

I'm thinking for a more general recursion fix we'll want to handle that on it's own because to handle it with minimal overhead, we'll likely want to discuss on it's own. Like I think we'll want a mix of strategy like what unroll custom definitions is doing with an equivalence library lookup and also checking `op.definition` to see if there are any known high level objects to synthesize in the definition.

Reply:
These are very good points. Based on this and other comments, in [1c71712](https://github.com/Qiskit/qiskit/pull/9846/commits/1c71712fec9a0b670e186b5d6dbb3f55594b2491) I am limiting the scope of this PR to avoid recursing into gates' definitions. That is, for now the annotated gates can only appear on top-level or as base gates of other annotated gates.